---
title: "Phylodynamic analysis of a COVID19 outbreak in Washington state"
author: "Josh Herbeck, modified from an original by Erik Volz"
date: "3/11/2020"
output: html_document
---

```{r knitr.setup, eval=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo=TRUE)
```

This script was originally written by Erik Volz (I cloned his https://github.com/emvolz/weifang-sarscov2 repository). I haved edited it so the anlaysis is of the 21 available WA SARS-CoV-2 sequences (as of March 10), not Weifang SARS-CoV-2 sequences. 
Update: there are more WA seqs available as of March 11.

### Load the R packages
```{r setup, warning=FALSE, message=FALSE}
library( ape ) 
library( lubridate ) 
library( treedater )
```

### Set the data directory and load the data
```{r sequences}
data_dir <- "~/Dropbox/SARS-CoV-2/data"
d <- read.dna(file.path(data_dir,'gisaid_cov2020_seqs_WA.plus.refs.fasta'), format = 'fasta')
raw <- dist.dna( d, model = 'raw' ) 
hky <- dist.dna( d, model = 'f84' )

.raw <- as.matrix( raw )
diag(.raw) <- NA 
print( mean( na.omit(as.vector(  .raw )) ) * 29289 ) 
#Erik had 29855 here; my alignment is a bit shorter so I use 29289 nucleotides.
#This estimates the average number of pairwise nucleotide differences found in the alignment
#3.27 was Erik's estimate from the Weifang alignment.

tr <- read.tree(file.path(data_dir,'gisaid_cov2020_seqs_WA.plus.refs_RAxML_Tree.newick'))
length(tr$tip.label)
```

We use the treedater algorithm and R package. This takes the following inputs:
  ```dater(tree, sts, s)```
  where 
* `tree` is an `ape::phylo` phylogeny object, 
* `sts` is a vector of sample times for each tip in `tree` formatted as fractions of the calendar year (e.g. 2010.1235)
* `s` is the length of the genetic sequences used to estimate `tree`

### Create the sts vector for treedater
```{r sts}
sts <- sapply( strsplit( tr$tip.label, '_' ), function(x) (x[length(x)] ) )
sts <- as.Date(sts)
sts <- decimal_date(sts)
names(sts) <- tr$tip.label
```

### Run treedater
```{r treedater, message=FALSE}
td <- dater( tr, sts, s = 29289 , searchRoot = 10, strictClock = FALSE, meanRateLimits = c(.0007, .0015) )
```

Treedater estimated tMRCA (time to most recent common ancestor) is 72 days, and the date of common ancestor as ```date_decimal(td[[4]])```.

### Plot tree
Washington sequences tip labels are in red font.
```{r}
#length(grep("USA/WA", tr$tip.label))
#tipcol <- rep('black', length(tr$tip.label))
#WA.tips <- c("USA/WA")
#colorsList <- c("red")
#for(i in 1:length(WA.tips)){tipcol[grep(WA.tips[i], tr$tip.label)] <- colorsList[i]}
#plot(tr, no.mar=T, tip.color = tipcol, cex=0.4)
plot( td, no.mar=T, cex = .4 )
```

### Plot root to trip regression
```{r}
rootToTipRegressionPlot(td, main="Root to tip regression", bty='n')
legend("topleft", 
       legend = c("internal nodes", "sample"), 
       col=c("black", "red"),
       pch=c(1,1),
       bty='n')
```

### Identify outliers
```{r}
outliers <- outlierTips( td , alpha = 0.2) 
```

```{r eval=FALSE}
pdf('Washington-treedater.pdf')
plot( td ); axisPhylo(root.time = 2020, backward=FALSE)
dev.off() 

pdf('Washington-rtt.pdf')
rootToTipRegressionPlot( td , bty='n')
dev.off() 
```

```{r eval=FALSE}
ot0 = outlierTips( td ) 

dd = read.dna( 'algn.21.1.fasta' , format ='fasta')
toremove <- c( 'Wuhan_HBCDC-HB-01_2019_2019.99452054795_exog', 'USA_CA1_2020_2020.06010928962_exog' )
dd <- dd[ setdiff( rownames(dd), toremove ), ]
write.dna(dd,  'algn.21.2.fasta', format = 'fasta' )
```
