---
title: "Phylodynamic analysis of a COVID19 outbreak in Washington state"
author: "Josh Herbeck, modified from an original by Erik Volz"
date: "3/20/2020"
output: html_document
---

```{r knitr.setup, eval=TRUE, echo=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=6)
```

<<<<<<< HEAD
This analysis was originally written by Erik Volz (I cloned his https://github.com/emvolz/weifang-sarscov2 repository). I haved edited it so the anlaysis is of the 69 available WA SARS-CoV-2 sequences (as of date March 19, 2020), not Weifang SARS-CoV-2 sequences. 
=======
This analysis was originally written by Erik Volz (I cloned his https://github.com/emvolz/weifang-sarscov2 repository). I have edited it so the anlaysis is of the X available WA SARS-CoV-2 sequences (as of date Y), not Weifang SARS-CoV-2 sequences. 
>>>>>>> bf764d485064ba0ec47d39d3d457e2d349fa9624

## Preliminary phylogenetic analysis

### Load the R packages
```{r setup}
library( ape ) 
library( lubridate ) 
library( treedater )
```

### Set the data directory and load the data

SARS-CoV-19 genomes are retrieved from: https://gisaid.org/CoV2020 

"Via the Nextstrain genome alignment pipeline, remove the first 130 sites, the last 50 sites, and site 18529 from the alignment."

```{r eval=TRUE, echo=FALSE}
prep_tip_labels_phydyn <- function( path_to_align, path_to_save = NULL
                                    , regexprs = c( '.*/WA/.*', '.*/WA/.*' ) 
                                    , invert_regexpr = c( FALSE, TRUE )
                                    , demes = c( 'Il'  , 'exog'  )
){

  if ( inherits( path_to_align, 'DNAbin' ) )
    d = path_to_align
  else
    d = read.dna( path_to_align, format = 'fasta')
  
  sids = rownames(d) 
  if ( length( regexprs ) != length( demes ))
    stop('Must provide equal numbers of regex and deme names ') 
  
  demegroups = lapply( 1:length(demes), function(k) {
    x = regexprs[k]
    if ( invert_regexpr[k] ) {
      return( sids[ !grepl( pattern = x , sids ) ] )
    }else{
      return( sids[ grepl( pattern = x , sids ) ] )
    }
  })
  int <- do.call( intersect, demegroups )
  uni = do.call( c, demegroups )
  if ( length( int ) > 0 )
    stop( 'Intersection of deme groups is non-empty. Each regex must match a unique set.' )
  if ( length( uni ) < length(sids) ){
    print( setdiff( sids, uni ))
    stop( 'There were some sequence IDs that did not match a regex. ' )
  }
  
  deme <- setNames( rep(demes[1], nrow(d)), sids )
  for ( k in 1:length( demes )){
    deme[ demegroups[[k]] ] <- demes[k]
  }
  
  sts <- sapply( strsplit( rownames(d), '\\|' ) , function(x){
    round(decimal_date( ymd( tail(x,1))), digits = 3)
  })
  rownames(d) <- paste(sep='|', rownames(d), sts, paste0('_', deme) )
  rownames(d) <- gsub( rownames(d), pattern = '\\s' , replacement = '_')
  row.names(d) <- gsub( rownames(d), pattern = "hCoV-19/", replacement = '')
  if ( !is.null( path_to_save ))
    write.dna( d, file = path_to_save, format = 'fasta' )
  d
}
```

```{r sequences}
data_dir <- "~/Dropbox/SARS-CoV-2/data"

d <- read.dna(file.path(data_dir,'gisaid.March18.alignment.edited.fasta'), format = 'fasta')

#Editing taxon names from GISAID format
#Erik's script to prep sequence names for BEAST PhyDyn runs
#I made slight edit to additionally remove "hCoV-19/" from taxon name
prep_tip_labels_phydyn(file.path(data_dir,'gisaid.March18.alignment.edited.fasta')
   , regexprs = c( '.*/WA/.*', '.*/WA/.*' ) 
	 , invert_regexpr = c( FALSE, TRUE )
	 , demes = c( 'Il'  , 'exog'  )
	 , path_to_save = file.path(data_dir,'edited.alignment.fasta') 
	)
```

```{r}
#Estimates the average number of pairwise nucleotide differences found in the alignment
d <- read.dna(file.path(data_dir,'edited.alignment.fasta'), format = 'fasta')

raw <- dist.dna( d, model = 'raw' ) 
hky <- dist.dna( d, model = 'f84' )

.raw <- as.matrix( raw )
diag(.raw) <- NA 
print( mean( na.omit(as.vector(  .raw )) ) * 29726 ) 
#Erik had 29855 here; my alignment is a bit shorter so I use 29726 nucleotides.
#This will get larger with more sequences as the epidemic progresses, obvs.
```


Loading a RAxML tree from Geneious (from the above alignment)
```{r}
tr <- read.tree(file.path(data_dir,'edited.alignment_RAxML_Tree.newick'))
length(tr$tip.label)
```


###Run RAxML to make an ML tree
```{r}
#alignment.raxml.tree <- raxml(d, m = "GTRGAMMA", f = "d", N = 1)
```

We use the treedater R package. This takes the following inputs:
  ```dater(tree, sts, s)```
  where 
* `tree` is an `ape::phylo` phylogeny object, 
* `sts` is a vector of sample times for each tip in `tree` formatted as fractions of the calendar year (e.g. 2010.1235)
* `s` is the length of the genetic sequences used to estimate `tree`

### Create the sts vector for treedater
```{r sts}
sts <- sapply( strsplit( tr$tip.label, '\\|' ), function(x) (x[length(x)-1] ) )
sts <- as.numeric(sts)
#sts <- decimal_date(sts)
names(sts) <- tr$tip.label

#"USA/WA-UW52/2020|EPI_ISL_415617|2020-03-09|2020.186|_exog"
```

### Run treedater
```{r treedater, message=FALSE}
td <- dater( tr, sts, s = 29726 , searchRoot = 10, temporalConstraints=TRUE,
             meanRateLimits = c(.0007, .002) )
```

Treedater estimated tMRCA (time to most recent common ancestor) is ```365*td$timeToMRCA``` days, and the date of common ancestor as ```date_decimal(td[[4]])```.

### Plot tree
Washington sequences tip labels are in red font. I need to change this so the names are not on the tips but rather just red and black dots or similar.
```{r}
length(grep("WA", tr$tip.label))
tipcol <- rep('black', length(tr$tip.label))
WA.tips <- c("WA")
colorsList <- c("red")
for(i in 1:length(WA.tips)){tipcol[grep(WA.tips[i], tr$tip.label)] <- colorsList[i]}
#plot(tr, no.mar=T, tip.color = tipcol, cex=0.3)
plot( td, no.mar=T, cex = .33 )
```

### Plot root to trip regression
```{r}
rootToTipRegressionPlot(td, main="Root to tip regression", bty='n')
legend("topleft", 
       legend = c("internal nodes", "sample"), 
       col=c("black", "red"),
       pch=c(1,1),
       bty='n')
```

### Identify and remove outliers (based on strict clock)
```{r}
ot0 <- outlierTips( td ) 
outliers.0 <- subset( ot0[ot0$q<0.01, ] )

dd <- d
#dd <- read.dna(file.path(data_dir,'gisaid.March18.alignment.edited.fasta'), format = 'fasta')
toremove <- as.character(outliers.0$taxon)
dd <- dd[ setdiff( rownames(dd), toremove ), ]
write.dna(dd,  'alignment.no.outliers.fasta', format = 'fasta' )
```






##BEAST phylodynamic analysis

Install BEAST2 from here: https://www.beast2.org/

###BEAST input files

https://github.com/mrc-ide/PhyDyn/wiki/Structured-Tree-Likelihood#population-structure
"Population structure is introduced by mapping tip nodes to deme names. This can be done by annotating the corresponding tree taxa ids with deme information or by introducing an extra type trait. With the former, taxa ids must be suffixed with an underscore followed by the deme information. For example, a tree tip annotated with taxon

  "EBOV|G4973.1|KR105286|SLE|x|2014-08-12_I"  
  
will be associated to deme I." 

In BEAUti:
  load the alignment
  assign tip dates
  choose the HKY+G substitution model and the strict clock model.

Useful template for implementing the SEI(J)R runs: https://github.com/mrc-ide/PhyDyn/wiki/SEIR-Template

https://beast.community/online_inference


###Processing of BEAST output files

Trees:

```{r eval=FALSE, echo=FALSE}
alignment.trees <- phytools::as.multiPhylo(file.path(data_dir,'alignment_trees'))
plot_densitree(
  file.path(data_dir,'alignment_trees'),
  alpha = 0.01,
  consensus = as.character(c(1:4)),
  cex = 2.0,
  scaleX = TRUE,
  scale.bar = FALSE
)
```


R0 estimation:

combineLogs0.R -> combinedLog.rds ->R0.R
